python/pyarrow/table.pxi

```diff
@@ -3401,6 +3401,9 @@ cdef class Table(_PandasConvertible):
                              .format(self.schema.names, target_schema.names))
 
         for column, field in zip(self.itercolumns(), target_schema):
+            if column.null_count > 0 and not field.nullable:
```

Comment:
```suggestion
            if not field.nullable and column.null_count > 0:
```

Switching the order will avoid checking the null_count (potentially expensive) if the field is nullable (which will be the most common case, since this is the default)

Reply:
Makes sense! Thanks!